source/extensions/access_loggers/filters/process_ratelimit/filter.cc

```diff
@@ -0,0 +1,62 @@
+#include "source/extensions/access_loggers/filters/process_ratelimit/filter.h"
+
+#include "envoy/access_log/access_log.h"
+
+#include "source/common/init/target_impl.h"
+#include "source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Filters {
+namespace ProcessRateLimit {
+
+ProcessRateLimitFilter::ProcessRateLimitFilter(
+    Server::Configuration::ServerFactoryContext& context,
+    const envoy::extensions::access_loggers::filters::process_ratelimit::v3::ProcessRateLimitFilter&
+        config)
+    : setter_key_(reinterpret_cast<intptr_t>(this)), cancel_cb_(std::make_shared<bool>(false)),
+      context_(context) {
+  auto setter =
+      [this, cancel_cb = std::shared_ptr<bool>(cancel_cb_)](
+          Envoy::Extensions::Filters::Common::LocalRateLimit::LocalRateLimiterSharedPtr limiter)
+      -> void {
+    if (*cancel_cb) {
+      return;
+    }
+    ENVOY_BUG(limiter != nullptr, "limiter shouldn't be null if the `limiter` is set from "
+                                  "callback.");
+
+    rate_limiter_->setLimiter(limiter);
+  };
+
+  if (!config.has_dynamic_config()) {
+    ExceptionUtil::throwEnvoyException("`dynamic_config` is required.");
+  }
+  rate_limiter_ = Envoy::Extensions::Filters::Common::LocalRateLimit::RateLimiterProviderSingleton::
+      getRateLimiter(context, config.dynamic_config().resource_name(),
+                     config.dynamic_config().config_source(), setter_key_, std::move(setter));
+}
+
+ProcessRateLimitFilter::~ProcessRateLimitFilter() {
+  *cancel_cb_ = true;
```

Comment:
I think this is not thread safe? Don't you want an atomic bool since you are writing here from a worker, but reading from the main?

Reply:
Updated. How about now?